File: leetcode/leet_22.py
# 22. Generate Parentheses
# Medium
# Topics
# Companies
# Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.


# Example 1:

# Input: n = 3
# Output: ["((()))","(()())","(())()","()(())","()()()"]
# Example 2:

# Input: n = 1
# Output: ["()"]


# Constraints:

# 1 <= n <= 8


from typing import List


def generateParenthesis(n: int) -> List[str]:
    lc = 0
    rc = 0
    ans = []
    s = ""
    count = 0

    def brack(lc, rc, string, count, type):
        nonlocal n
        print("\ncount: ", count, end=" |  ")
        print("type : ", type, end="  |  ")
        print("string : ", string, end=f"{(10 - len(string))* " "}|  ")
        print("lc : ", lc, end="  |  ")
        print("rc : ", rc, end="  |  ")

        if lc == n and rc == n:
            ans.append(string)
            print(f"APPENDED - {string}")
            return
        if lc <= n and rc <= n:
            # string += "("
            # lc += 1
            brack(lc + 1, rc, string + "(", count + 1, "left ")

        if rc < lc:
            # string += ")"
            # rc += 1
            brack(lc, rc + 1, string + ")", count + 1, "right")

    brack(lc, rc, s, 0, "START")
    return ans


if __name__ == "__main__":
    print(generateParenthesis(3))

File: leetcode/leet_22.cpp
// # 22. Generate Parentheses
// # Medium
// # Topics
// # Companies
// # Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

// # Example 1:

// # Input: n = 3
// # Output: ["((()))","(()())","(())()","()(())","()()()"]
// # Example 2:

// # Input: n = 1
// # Output: ["()"]

// # Constraints:

// # 1 <= n <= 8
#include <string>
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

class Solution
{
private:
    vector<string> ans;
    void brack(string str, int l, int r, int n)
    {

        if (l == n && r == n)
        {
            ans.push_back(str);
            return;
        }

        if (l < n)
        {
            brack(str + '(', l + 1, r, n);
        }
        if (r < l)
        {
            brack(str + ')', l, r + 1, n);
        }
    }

public:
    vector<string> generateParenthesis(int n)
    {
        int l = 0;
        int r = 0;
        brack("", l, r, n);
        return ans;
    }
};

int main()
{
    Solution S;
    vector<string> v;
    v = S.generateParenthesis(3);
    for (const auto &str : v)
    {
        cout << str << endl; // Print each string on a new line
    }
}